src/GridFS/GridFSAdapter.php

```
@@ -0,0 +1,413 @@
+    }
+
+    public function writeStream(string $path, $contents, Config $config): void
+    {
+        if (str_ends_with($path, '/')) {
+            throw UnableToWriteFile::atLocation($path, 'file path cannot end with a slash');
+        }
+
+        $filename = $this->prefixer->prefixPath($path);
+        $options = [];
+        if ($visibility = $config->get(Config::OPTION_VISIBILITY)) {
+            $options['metadata'][self::METADATA_VISIBILITY] = $visibility;
+        }
+        if (($mimetype = $config->get('mimetype')) || ($mimetype = $this->mimeTypeDetector->detectMimeTypeFromPath($path))) {
+            $options['metadata'][self::METADATA_MIMETYPE] = $mimetype;
+        }
+
+        try {
+            $this->bucket->uploadFromStream($filename, $contents, $options);
+        } catch (Exception $exception) {
+            throw UnableToWriteFile::atLocation($path, $exception->getMessage(), $exception);
+        }
+    }
+
+    public function read(string $path): string
+    {
+        return stream_get_contents($this->readStream($path));
+    }
+
+    public function readStream(string $path)
+    {
+        if (str_ends_with($path, '/')) {
+            throw UnableToReadFile::fromLocation($path, 'file path cannot end with a slash');
+        }
+
+        $file = $this->findFile($path);
+
+        if ($file === null) {
+            throw UnableToReadFile::fromLocation($path, 'file does not exist');
+        }
+
+        try {
+            return $this->bucket->openDownloadStream($file['_id']);
+        } catch (Exception $exception) {
+            throw UnableToReadFile::fromLocation($path, $exception->getMessage(), $exception);
+        }
+    }
+
+    public function delete(string $path): void
+    {
+        if (str_ends_with($path, '/')) {
+            throw UnableToDeleteFile::atLocation($path, 'file path cannot end with a slash');
+        }
+
+        // Deleting a file that does not exist is no-op
+        while ($file = $this->findFile($path)) {
```

1) why is this a while loop?
2) this block may be affected by a race condition when two deletions are happening concurrently. In other implementations we try to delete and ignore the exception if it was caused by a "file not found" situation. Can we inspect the exception to see if occurs and ignore the error when that's the case?

---

The while was necessary to remove all the revisions of the file name. I replaced the code with a "find all revisions" followed by "delete" on each revision. Good point about race-condition, I added a try-catch.